fastparquet/writer.py

```diff
@@ -944,18 +1019,60 @@ def write(filename, data, row_group_offsets=50000000,
                 raise ValueError('Requested file scheme is %s, but '
                                  'existing file scheme is not.' % file_scheme)
             fmd = pf.fmd
-            i_offset = find_max_part(fmd.row_groups)
+            if rgp_by_date:
+                if date_col:
+                    # Use date column provided.
+                    date_index_name = date_col
+                else:
+                    # Date index is 1st column
+                    date_index_name = index_cols[0]
+
+                # Formatting `drop_duplicates_on` for use in `pd.concat()`.
+                if drop_duplicates_on:
+                    if isinstance(drop_duplicates_on, str):
+                        # Case `drop_duplicates_on` is a single column name.
+                        drop_duplicates_on = [drop_duplicates_on]
+                    if not date_index_name in drop_duplicates_on:
+                        drop_duplicates_on.append(date_index_name)
+                    # Check all values in `drop_duplicates_on` exist
+                    # in column list.
+                    if not all(item in pf.columns for item in drop_duplicates_on):
+                        raise ValueError('At least one value in {!s} is not a \
+column/index name as in {!s}.'.format(str(drop_duplicates_on),
+                                      str(list(pf.columns))))
+                # Retrieving list of existing parquet part number. It is used
+                # to determine position of updated/added row group in
+                # `ParquetFile.row_groups` (list of row groups).
+                # Dict maintains order. This feature is used to keep track
+                # where to insert or replace row groups appropriately in the list.
+                part_number_to_rgp = {\
+        int(rg.columns[0].file_path.split('.')[1]): rg for rg in pf.row_groups}
+                part_numbers = list(part_number_to_rgp.keys())
+            else:
+                i_offset = find_max_part(fmd.row_groups)
+
             if tuple(partition_on) != tuple(pf.cats):
                 raise ValueError('When appending, partitioning columns must'
                                  ' match existing data')
         else:
             i_offset = 0
-        fn = join_path(filename, '_metadata')
+
         mkdirs(filename)
+
         for i, start in enumerate(row_group_offsets):
-            end = (row_group_offsets[i+1] if i < (len(row_group_offsets) - 1)
-                   else None)
-            part = 'part.%i.parquet' % (i + i_offset)
+            if rgp_by_date:
+                # Get `end` before `start` pd.Period is then overwritten as
+                # a pd.Timestamp
+                end = start.end_time
```

Comment:
> Is end always the same as the start of the next period?

No. Here, `start`object if a `pandas.Period`.
You can see later when I use the boundaries of this perido, I use following inequalities.
[` data_slice = data.loc[(date_index >= start) & (date_index <= end)]`](https://github.com/yohplala/fastparquet/blob/46f3982a5c722de2bf707e72416661370d9fbaa8/fastparquet/writer.py#L1089)

Using strict inequality for upper bound makes it possible to use the start of the next period.
But then, I need
- to access this next period: it makes the loop a bit more complex
- I need to take care to generate one Period more in the end of the list, so that the last element of the Period list is not lying withing the data spanned by the dataset that is being recorded.
This can be done, but is more complex than simply calling `end_time` on a Period I already have in my hand.

Reply:
> The column statistics max/min values should align with these, right,

No. The column statistics have min/max for existing data in the dataset. But if you have a time span with 'hole' in this time span?
`pandas.date_range` generates regular time span (period)